runtime/sema/check_assignment.go

```diff
@@ -107,6 +110,83 @@ func (checker *Checker) checkAssignment(
 	return
 }
 
+func (checker *Checker) enforcePureAssignment(assignment ast.Statement, target ast.Expression) {
+	if !checker.CurrentPurityScope().EnforcePurity {
+		return
+	}
+
+	var variable *Variable
+
+	switch targetExp := target.(type) {
+	case *ast.IdentifierExpression:
+		variable = checker.valueActivations.Find(targetExp.Identifier.Identifier)
+	case *ast.IndexExpression:
+		if indexIdentifier, ok := targetExp.TargetExpression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	case *ast.MemberExpression:
+		if indexIdentifier, ok := targetExp.Expression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	}
+
+	// if the target is not a variable (e.g. a nested index expression x[0][0], or a nested
+	// member expression x.y.z), the analysis cannot know for sure where the value being
+	// assigned to originated, so we must default to impure. Consider:
+	// -----------------------------------------------------------------------------
+	// let a: &[Int] = &[0]
+	// pure fun foo(_ x: Int) {
+	//     let b: &[Int] = &[0]
+	//     let c = [a, b]
+	//     c[x][0] = 4 // we cannot know statically whether a or b receives the write here
+	// }
```

Comment:
Hmm,  I think you're right, although we also need to check whether any accessed member is a resource as well; since those are moved rather than copied.